### DISCUSSION TITLE
Multiarch builds

### DISCUSSION THREAD

**Discussion by modem7** [gboudreau/Greyhole#279] [Show and tell]
Heya,

I've just been doing some work with multi-arch builds on Docker using DroneCI, and decided to try building Greyhole to see if it'd work, so it'd work on things like a Pi or latest ARM based Mac.

Unsure if a multi-arch auto-build system would be useful, I'd be happy to discuss further if it's useful for you.

https://hub.docker.com/repository/docker/modem7/samba-greyhole
https://drone.modem7.com/modem7/samba-greyhole-docker/10

Can happily discuss further either here or elsewhere (so not to clog up issues w/ off-topic chatter) if it's useful!

**Comment by gboudreau** [comment]
Hi,

I was already creating the `x86_64` docker image with my build script (i.e. at the same time I was creating RPM/DEB packages, and uploading a new release on Github), so [I just changed](https://github.com/gboudreau/Greyhole/commit/44efeb2fba0e25f140c86bca548249e481bc89fc) that script to create multi-arch images.

Also, Docker Hub disabled auto-builds a few months ago, for free users, so the `develop` tags on Docker stopped being up to date then. I disabled that integration on Github, and replaced it with a webhook that simply pokes my server when a new push is made, and that now triggers a new Docker multi-arch build that will be tagged `develop`, and pushed to Docker Hub.

The only thing that is still missing there is that the [`install_greyhole_vfs.sh`](https://github.com/gboudreau/samba-greyhole-docker/blob/master/install_greyhole_vfs.sh) script, used in the [`Dockerfile`](https://github.com/gboudreau/samba-greyhole-docker/blob/master/Dockerfile), is NOT aware of the Docker image platform, and thus is trying to use the pre-compiled VFS module for `x86_64` on `arm` platforms.
I made a change to fix that, but now that it is trying to build the VFS in `arm` images, the compilation fails. I'm investigating that, and will commit the fix when I find something, and will also publish the pre-compiled VFS modules for `arm` on my server, to speed up future Docker builds.

Cheers.
(And thanks for the idea for multi-arch build.)

**Comment by modem7** [comment]
Another note I found (from [Hadolint](https://github.com/hadolint/hadolint)) is:

Dockerfile:3 DL3018 style: Pin versions in apk add. Instead of `apk add <package>` use `apk add <package>=<version>`

Dockerfile:11 SC2006 style: Use $(...) notation instead of legacy backticked `...`.

Dockerfile:15 DL3020 style: Use COPY instead of ADD for files and folders

Dockerfile:16 DL3003 style: Use WORKDIR to switch to a directory

Doesn't change how it works obviously, but more something to keep in mind if it bugs you.

**Comment by gboudreau** [comment]
Thanks for those.
L3: Since I use a fixed alpine version in `FROM` (L1), packages don't change (enough) for that to matter.
L11: Yes, indeed. I also prefer `$()`.
L15: I need to use `ADD`, since `COPY` doesn't support URLs.
L16: I don't want to change the `WORKDIR` permanently; I do that on L65. I would need to split my `RUN` instruction in two, to be able to use `WORKDIR` here, since the folder in question doesn't exist before it is downloaded and extracted by the few first lines on `RUN`.
